src/transformers/generation_utils.py

```diff
@@ -359,12 +359,72 @@ class BeamSampleEncoderDecoderOutput(ModelOutput):
 BeamSampleOutput = Union[BeamSampleEncoderDecoderOutput, BeamSampleDecoderOnlyOutput]
 
 
+ENCODER_MODEL_INPUT_NAMES = ["input_ids", "inputs_embeds", "input_values", "input_features", "pixel_values"]
+
+
 class GenerationMixin:
     """
     A class containing all of the functions supporting generation, to be used as a mixin in
     :class:`~transformers.PreTrainedModel`.
     """
 
+    def _prepare_model_inputs(
+        self,
+        inputs: Optional[torch.Tensor] = None,
+        bos_token_id: Optional[int] = None,
+        model_kwargs: Optional[Dict[str, torch.Tensor]] = None,
+    ) -> Tuple[torch.Tensor, Optional[str]]:
+        """
+        This function extracts the model-specific `inputs` for generation.
+        """
+        # filter model input names that are `None`
+        model_kwargs = {k: v for k, v in model_kwargs.items() if k not in ENCODER_MODEL_INPUT_NAMES or v is not None}
+        # extract keyword arguments that are model input specific
+        model_input_kwarg_names = set(ENCODER_MODEL_INPUT_NAMES) & set(model_kwargs.keys())
+
+        # There are 5 possible scenarios
+        if inputs is not None and len(model_input_kwarg_names) == 0:
+            # 1. `inputs` are passed and no model-specific keyword inputs
+            # -> return input
+            model_input_name = None
+            return inputs, model_input_name, model_kwargs
+        elif inputs is not None and len(model_input_kwarg_names) > 0:
+            # 2. `inputs` are passed as well as model-specific keyword inputs
+            # -> not allowed, raise Error
+            raise ValueError(
+                f"`inputs`: {inputs}` were passed alongside "
+                f"{model_input_kwarg_names} which is not allowed."
+                f"Make sure to not pass any of {model_input_kwarg_names} "
+                "when `inputs` is defined."
+            )
+        elif inputs is None and len(model_input_kwarg_names) == 0:
+            # 3. no `inputs` and no model-specific keyword inputs are passed
+            # -> try to create `input_ids` from BOS
+            input_tensor = self._prepare_input_ids_for_generation(bos_token_id, model_kwargs.get("encoder_outputs"))
+            return input_tensor, "input_ids", model_kwargs
+        elif inputs is None and len(model_input_kwarg_names) == 1:
+            # 4. no `inputs` are passed and exactly one model-specific keyword input
+            # -> return that model-specific keyword input tensor
+            model_input_name = model_input_kwarg_names.pop()
+            input_tensor = model_kwargs.pop(model_input_name)
+
+            # make sure model is encoder decoder if not `input_ids`
+            if not self.config.is_encoder_decoder and model_input_name != "input_ids":
```

Comment:
This breaks some image models:
```py
output = model.generate(pixel_values=context, max_length=model.config.n_positions + 1, temperature=1.0, top_k=40)
```

with ImageGPT returns
```
ValueError: If pixel_values is passed as model-specific keyword input then model has to be an encoder-decoder and not a ImageGPTForCausalImageModeling.
```

Whereas
```py
output = model.generate(context, max_length=model.config.n_positions + 1, temperature=1.0, top_k=40)
```
works

Is this expected as ImageGPT is not a `VisionEncoderDecoderModel` ?

Reply:
Great catch! Didn't know that we already have a multi-modal decoder-only model

Reply:
Actually looking a bit more into in - I'm wondering whether `pixel_values` was the correct name for ImageGPT's input. According to the docs: https://huggingface.co/docs/transformers/master/en/model_doc/imagegpt#transformers.ImageGPTModel.forward.pixel_values

The input is of shape `(batch_size, sequence_length)` and the type is `torch.long` which is exactly what `input_ids` look like and not really `pixel_values`.

@LysandreJik @NielsRogge @sgugger - I don't know image-gpt too well, but it looks a bit like to me that the better option here is to deprecate the `pixel_values` input for ImageGPT. In case you agree, I think we can go ahead with this PR as it doesn't break anything that previously worked and making ImageGPT work can be done in a future PR. 

If not and you think the better name is indeed `pixel_values` than I need to change the logic here a bit again.

Reply:
Yes agreed that it's not the best name for this model.

Reply:
Fine for you guys to deprecate and rename it?

Reply:
In a follow-up PR

Reply:
Yes!